src/deno.h

```diff
@@ -3,36 +3,84 @@
+#include <stdio.h>
+#include <stdlib.h>
+
 // Neither Rust nor Go support calling directly into C++ functions, therefore
 // the public interface to libdeno is done in C.
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 // Data that gets transmitted.
-typedef struct {
-  uint8_t* alloc_ptr;  // Start of memory allocation (returned from `malloc()`).
-  size_t alloc_len;    // Length of the memory allocation.
-  uint8_t* data_ptr;   // Start of logical contents (within the allocation).
-  size_t data_len;     // Length of logical contents.
+typedef struct deno_buf {
+  uint8_t* const alloc_ptr;  // Start of memory allocation. Read-only.
+  const size_t alloc_len;    // Length of the memory allocation. Read-only.
+  uint8_t* data_ptr;  // Start of logical contents (within the allocation).
+  size_t data_len;    // Length of logical contents.
+
+#ifdef __cplusplus
+  // Since deno_buf has a C abi, we can't rely on the destructor to run, so
+  // the user has to call deno_buf_delete() explicitly. However, if we are in
+  // C++, use the destructor to catch bugs.
+  ~deno_buf();
+  // Disallow pass-by-value.
+  deno_buf(deno_buf&) = delete;
+  deno_buf(const deno_buf&) = default;
+
+#endif
 } deno_buf;
 
+// Clang will complain that the deno_buf_* functions have C linkage while
+// `struct deno_buf` has a destructor. This is intentional -- suppress the
+// warning.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wreturn-type-c-linkage"
+
+static const deno_buf DENO_BUF_INIT = {nullptr, 0, nullptr, 0};
+
+// Allocate a new buffer with backing storage.
+const deno_buf deno_buf_new(size_t len, bool zero_init);
+// Allocate new storage for an existing buffer. The `data_ptr` and `data_len`
+// fields are initialized to use the entire allocation.
+void deno_buf_delete(deno_buf* buf);
+// Move ownership of a buffer. Returns a new buffer with the same contents as
+// the source buffer. Afterwards, the source buffer is reset to null.
+const deno_buf deno_buf_move(deno_buf* source_buf);
+void deno_buf_move_into(deno_buf* target_buf, deno_buf* source_buf);
+// Returns true if `buf` is a null buffer.
+bool deno_buf_is_null(const deno_buf* buf);
+
+// Directly manipulate deno_buf fields, without doing any memory management.
+// This is for special use cases - don't use this unless unavoidable.
+const deno_buf deno_buf_new_raw(uint8_t* alloc_ptr, size_t alloc_len,
+                                uint8_t* data_ptr, size_t data_len);
+void deno_buf_delete_raw(deno_buf* buf);
+
+#pragma clang diagnostic pop
+
 struct deno_s;
 typedef struct deno_s Deno;
 
 // A callback to receive a message from deno.send javascript call.
 // buf is valid only for the lifetime of the call.
-typedef void (*deno_recv_cb)(Deno* d, deno_buf buf);
+typedef void (*deno_recv_cb)(Deno* d, deno_buf* buf);
+
+// This callback receives a deno_buf containing a message. It should return a
+// message's cmd_id.
+typedef uint32_t (*deno_cmd_id_cb)(const deno_buf* buf);
```

Comment:
This doesn't seem like the most elegant design. If cmd_id is going to be required for the usage of libdeno, then it should probably be a parameter in deno_send and deno_recv_cb.